Documentation/contributor-guide/features.md

```diff
@@ -0,0 +1,39 @@
+# Features 
+
+This document provides general guidelines for adding and removing etcd features. The etcd features fall into two stages, experimental and stable.
+
+## Experimental
+
+Any new feature should be added as an experimental feature. An experimental feature is characterized as below:
+- Associated with an issue with a clear need for such a feature.
+- Any configuration flags related to the implementation of the feature are prefixed with `experimental` e.g. `--experimental-feature-name`.
+- Any variable names related to the implementation of the feature are prefixed with `Experimental` e.g. `ExperimentalFeatureName`.
+- Might be buggy. Enabling the feature may not work as expected. 
+- Support for such a feature may be dropped at any time without notice.
+- Disabled by default when added initially.
+
+### Graduation to Stable
```

Comment:
I expect that during graduation we will remove `experimental` prefix from feature flag. This would be a breaking change in flags that could be easily mitigated by not renaming old experimental flag, but by marking it as hidden and adding new one. There should not be a difference in behavior but it should bring big improvements to user experience during upgrade.

Reply:
I am thinking can we follow the same approach as [Kubernetes](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) to manage the feature?

Reply:
Feature gates are good when you don't need to do any configuration for feature to work. On multiple occasions I have added a new feature to Kubernetes and decided against feature gating. 

What I am proposing is more inline with https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli. From CLI perspective graduating a feature is a breaking change without any deprecation. We remove experimental flag and add a new one. During upgrade admins need to take a immediate action and update all graduated flags. 

I'm proposing that when feature graduates we still leave `experimental` flags for one release, giving users time to adapt.

Alternative approaches:
* Stop using `experimental` prefix and require user to opt in into experimental feature by providing separate `--experimental-feature-gate` flag like https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

Reply:
> * Stop using `experimental` prefix and require user to opt in into experimental feature by providing separate `--experimental-feature-gate` flag like https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

This is exactly what I am thinking. 

- Adding/removing the prefix `experimental` is indeed a breaking change; 
- Keeping `experimental` flags for one more release complicates the feature graduation process.

Reply:
Yes, good points both about breaking changes related to prefix, thanks. May be simple way would be to treat experimental features naming/usage in a similar way to unsafe features (I missed it in the stages earlier by we have stable, experimental and unsafe features currently). I will be pinging you for quick discussion around it.

Reply:
I think there are 3 separate things to address:
* Have documentation reflect the current state (usage of `experimental` prefix)
* Avoiding breaking changes in flags (not immidietly removing experimental flags upon graduation)
* Long term approach for handling feature flags that avoid problem of flag compatibility altogether. 

In this PR we should address first two points. @spzala feel free to open an issue with proposal for the third one.